std/assembly/util/string.ts

```diff
@@ -160,54 +180,229 @@ export function strtol<T>(str: string, radix: i32 = 0): T {
+  var pointed = false;
+  var consumed = 0;
+  var position = 0;
+  var x: u64 = 0;
+  if (code == CharCode.DOT) {
+    let noDigits = !(savedPtr - ptr);
+    ptr += 2; --len;
+    if (!len && noDigits) return NaN;
+    for (pointed = true; (code = <u32>load<u16>(ptr)) == CharCode._0; --position, ptr += 2) --len;
+    if (len <= 0) return 0;
+    if (!position && noDigits && code - CharCode._0 >= 10) return NaN;
+  }
+  for (let digit = code - CharCode._0; digit < 10 || (code == CharCode.DOT && !pointed); digit = code - CharCode._0) {
+    if (digit < 10) {
+      x = consumed < capacity ? 10 * x + digit : x | u64(!!digit);
+      ++consumed;
+    } else {
+      position = consumed;
+      pointed = true;
+    }
+    if (!--len) break;
+    code = <u32>load<u16>(ptr += 2);
+  }
+
+  if (!pointed) position = consumed;
+  return copysign<f64>(scientific(x, position - min(capacity, consumed) + parseExp(ptr, len)), sign);
+}
+
+// @ts-ignore: decorator
+@inline
+function scientific(significand: u64, exp: i32): f64 {
+  if (!significand || exp < -342) return 0;
+  if (exp > 308) return Infinity;
+  // Try use fast path
+  var result = strtodFast(<f64>significand, exp);
+  if (!isNaN(result)) return result;
+  if (exp < 0) {
+    return scaledown(significand, exp);
+  } else {
+    return scaleup(significand, exp);
+  }
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h#L43
+// @ts-ignore: decorator
+@inline
+function scaledown(significand: u64, exp: i32): f64 {
+  const denom: u64 = 6103515625; // 1e14 * 0x1p-14
+  const scale = reinterpret<f64>(0x3F06849B86A12B9B); // 1e-14 * 0x1p32
+
+  var shift = clz(significand);
+  significand <<= shift;
+  shift = exp - shift;
+
+  for (; exp <= -14; exp += 14) {
+    let q = significand / denom;
+    let r = significand % denom;
+    let s = clz(q);
+    significand = (q << s) + <u64>nearest(scale * <f64>(r << (s - 18)));
+    shift -= s;
+  }
+  var b = <u64>ipow32(5, -exp);
+  var q = significand / b;
+  var r = significand % b;
+  var s = clz(q);
+  significand = (q << s) + <u64>(reinterpret<f64>(reinterpret<u64>(<f64>r) + (s << 52)) / <f64>b);
+  shift -= s;
+
+  return NativeMath.scalbn(<f64>significand, <i32>shift);
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h#L27
+// @ts-ignore: decorator
+@inline
+function scaleup(significand: u64, exp: i32): f64 {
+  const coeff: u32 = 1220703125; // 1e13 * 0x1p-13;
+  var shift = ctz(significand);
+  significand >>= shift;
+  shift += exp;
+
+  __fixmulShift = shift;
+  for (; exp >= 13; exp -= 13) {
+    significand = fixmul(significand, coeff);
+  }
+  significand = fixmul(significand, <u32>ipow32(5, exp));
+  shift = __fixmulShift;
+  return NativeMath.scalbn(<f64>significand, <i32>shift);
+}
+
+// Adopted from metallic lib:
+// https://github.com/jdh8/metallic/blob/master/src/stdlib/parse/scientific.h#L27
```

Comment:
Seems this reference should be `src/stdlib/parse/float.h#L8`. Also wondering if the line numbers can lead to more confusion than being helpful when metallic source code updates.

Reply:
Yeah, line details are unnecessary